sa/model.go

```diff
@@ -445,6 +446,72 @@ type authzModel struct {
 	ValidationRecord []byte     `db:"validationRecord"`
 }
 
+func eraseSliceType(input []string) []any {
+	var output []any
+	for _, in := range input {
+		output = append(output, in)
+	}
+	return output
+}
+
+// SelectAuthzsMatchingIssuance looks for a set of authzs that would have
+// authorized a given issuance that is known to have occurred. The returned
+// authzs will all belong to the given regID, will have potentially been valid
+// at the time of issuance, and will have the appropriate identifier type and
+// value. This may return multiple authzs for the same identifier type and value.
+//
+// This returns "potentially" valid authzs because a client may have set an
+// authzs status to deactivated after issuance, so we return both valid and
+// deactivated authzs. It also uses a small amount of leeway (1s) to account
+// for possible clock skew.
+//
+// This function doesn't do anything special for authzs with an expiration in
+// the past. If the stored authz has a valid status, it is returned with a
+// valid status regardless of whether it is also expired.
+func SelectAuthzsMatchingIssuance(
+	s db.Selector,
+	regID int64,
+	issued time.Time,
+	dnsNames []string,
+) ([]*corepb.Authorization, error) {
+	var authzModels []authzModel
+	query := fmt.Sprintf(`SELECT %s FROM authz2 WHERE
+			registrationID = ? AND
+			status IN (?, ?) AND
+			expires >= ? AND
+			attemptedAt <= ? AND
+			identifierType = ? AND
+			identifierValue IN (%s)`,
+		authzFields,
+		db.QuestionMarks(len(dnsNames)))
+	var args []any
+	args = append(args,
+		regID,
+		statusToUint[core.StatusValid],
+		statusToUint[core.StatusDeactivated],
+		issued.Add(-1*time.Second), // leeway for clock skew
+		issued.Add(1*time.Second),  // leeway for clock skew
+		identifierTypeToUint[string(identifier.DNS)],
+	)
+	args = append(args, eraseSliceType(dnsNames)...)
```

Comment:
Wow, it's wild to me that this eraseSliceType function is necessary. But is allocating a new slice and putting all the elements into it ourselves really more efficient than
```suggestion
	for _, n := range dnsNames {
		args = append(args, n)
	}
```
It's the same number of iterations though a loop, it's probably the same number of allocations (although eraseSliceType could maybe fix that by doing `output := make([]any, len(input))`), and it doesn't require the reader to go "wait why is this necessary?".

Reply:
I had a vague idea that `eraseSliceType` might be useful in other places in the SA, but I'll save that for a potential separate refactoring.